### PR TITLE
[Transform] Double Radius Node Labeling

### PR DESCRIPTION
**Pull request by rudongyu**
## Description

Add Double Radius Node Labeling transform from [SEAL](https://arxiv.org/abs/1802.09691) into dgl core as planned in #4372. 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by Rhett-Ying** [comment]
To trigger regression tests: 
 - `@dgl-bot run [instance-type] [which tests] [compare-with-branch]`; 
 For example: `@dgl-bot run g4dn.4xlarge all dmlc/master` or `@dgl-bot run c5.9xlarge kernel,api dmlc/master`

**Comment by Rhett-Ying** [comment]
Commit ID: e0c79f4b840d8c0d2e45f13d237c9dcbc4bd9f25

Build ID: 1

Status: ❌ CI test failed in Stage [Lint Check]. 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/1/1/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/1/1/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: b4038dc281e7bf3f32c45d23064f4e231051e121

Build ID: 2

Status: ❌ CI test failed in Stage [Lint Check]. 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/2/2/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/2/2/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: 367097ecfc86485fd366c4ecf77a2deb3debb473

Build ID: 3

Status: ❌ CI test failed in Stage [Torch CPU (Win64) Unit test]. 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/3/3/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/3/3/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: 7267d3e4a6693af8f4b6b3f6e0e522139505ae7d

Build ID: 4

Status: ❌ CI test failed in Stage [Torch CPU (Win64) Unit test]. 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/4/4/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/4/4/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: dedaf84f48096c6dec2085992e6a52cfca873b4d

Build ID: 5

Status: ❌ CI test failed in Stage [Torch CPU (Win64) Unit test]. 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/5/5/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/5/5/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: cbf90b6c50997a68379bca658b0006c81d3d423e

Build ID: 6

Status: ❌ CI test failed in Stage [Torch CPU (Win64) Unit test]. 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/6/6/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4513/6/6/logs/cireport.log)

File: python/dgl/transforms/functional.py [hunk 1]
```diff
@@ -3693,4 +3694,68 @@ def to_double(g):
...
+
+    where :math:`d_s` and :math:`d_t` denote the shortest distance to the source and
+    the target, respectively. :math:`d = d_s + d_t`.
+
+    Parameters
+    ----------
+    g : DGLGraph
+        The input graph.
+    src : int
+        The source node id of the target link.
```
Comment (mufeili) [inline python/dgl/transforms/functional.py:3718]:
```suggestion
        The source node ID of the target link.
```

for consistency with the rest APIs

File: python/dgl/transforms/functional.py [hunk 2]
```diff
@@ -3693,4 +3694,68 @@ def to_double(g):
...
+    dst : int
+        The destination node id of the target link.
```
Comment (mufeili) [inline python/dgl/transforms/functional.py:3720]: Ditto

File: python/dgl/transforms/functional.py [hunk 3]
```diff
@@ -3693,4 +3694,68 @@ def to_double(g):
...
+
+    Returns
+    -------
+    Tensor
+        The double radius node labeling Tensor of shape :math:`(N,)`, where
+        :math:`N` is the number of nodes in the input graph.
+
+    Example
+    -------
+    >>> import dgl
```
Comment (mufeili) [inline python/dgl/transforms/functional.py:3730]: Add an empty line above

File: python/dgl/transforms/functional.py [hunk 4]
```diff
@@ -3693,4 +3694,68 @@ def to_double(g):
...
+    idx = list(range(src)) + list(range(src + 1, adj.shape[0]))
+    adj_wo_src = adj[idx, :][:, idx]
+
+    idx = list(range(dst)) + list(range(dst + 1, adj.shape[0]))
+    adj_wo_dst = adj[idx, :][:, idx]
+
+    # distance to the source node
+    ds = sparse.csgraph.shortest_path(adj_wo_dst, directed=False, unweighted=True, indices=src)
+    ds = np.insert(ds, dst, 0, axis=0)
+    # distance to the destination node
```
Comment (mufeili) [inline python/dgl/transforms/functional.py:3747]: Ditto

File: python/dgl/transforms/functional.py [hunk 5]
```diff
@@ -3693,4 +3694,68 @@ def to_double(g):
...
+    dt = sparse.csgraph.shortest_path(adj_wo_src, directed=False, unweighted=True, indices=dst-1)
```
Comment (mufeili) [inline python/dgl/transforms/functional.py:3748]: Why `dst - 1` rather than `dst`?

File: python/dgl/transforms/functional.py [hunk 6]
```diff
@@ -3693,4 +3694,68 @@ def to_double(g):
...
+    dt = np.insert(dt, src, 0, axis=0)
+
+    d = ds + dt
+    # suppress invalid value (nan) warnings
+    with np.errstate(invalid='ignore'):
+        z = 1 + np.stack([ds, dt]).min(axis=0) + d//2 * (d//2 + d%2 -1)
```
Comment (mufeili) [inline python/dgl/transforms/functional.py:3754]:
```suggestion
        z = 1 + np.stack([ds, dt]).min(axis=0) + d//2 * (d//2 + d%2 - 1)
```

File: python/dgl/transforms/module.py [hunk 7]
```diff
@@ -403,6 +404,36 @@ def __call__(self, g):
 
         return g
 
+class DoubleRadiusNodeLabeling(BaseTransform):
```
Comment (mufeili) [inline python/dgl/transforms/module.py:407]: For this item, perhaps it suffices to have the functional interface without a module interface. Module interface helps for transforming multiple graphs in the same way when users provide a transform to a dataset class. However, in this case users need to additionally provide `src` and `dst`. So it's pretty much the same as a functional interface.

**Comment by mufeili** [comment]
Done a first pass

File: python/dgl/transforms/functional.py (hunk 5 above)
Comment (rudongyu) [reply python/dgl/transforms/functional.py:3846]: After removing src (< dst) in adj_wo_src, the new index of dst is dst - 1.

File: python/dgl/transforms/module.py (hunk 7 above)
Comment (rudongyu) [reply python/dgl/transforms/module.py]: agree. removed the module interface.

**Comment by rudongyu** [comment]
@mufeili addressed the issues in the 1st round review. Please check if you have any further comments.

File: python/dgl/transforms/functional.py (hunk 5 above)
Comment (mufeili) [reply python/dgl/transforms/functional.py:3846]: I see. Thanks.

**Review by mufeili — APPROVED** [review]
LGTM

**Pull request merged**